src/core/thread/mle_router.cpp

```diff
@@ -1106,6 +1107,13 @@ ThreadError MleRouter::HandleAdvertisement(const Message &aMessage, const Ip6::M
                      leaderData.GetWeighting(), peerParitionId,
                      mLeaderData.GetWeighting(), mLeaderData.GetPartitionId());
 
+        if (((mDeviceMode & ModeTlv::kModeFFD) == 0) && (memcmp(&mParent.mMacAddr, &macAddr, sizeof(router->mMacAddr)) == 0) &&
+            (mParent.mValid.mRloc16 == sourceAddress.GetRloc16()))
+        {
+            SetLeaderData(leaderData.GetPartitionId(), leaderData.GetWeighting(), leaderData.GetLeaderRouterId());
+            ExitNow();
+        }
```

Comment:
Hi @jwhui , I was thinking putting it here could also be executed for all Thread devices as long as it is not in detach state. And another benefit is it could reuse some code here to avoid some duplication. 
what do you think?